components/script/document_collection.rs

```diff
@@ -0,0 +1,177 @@
+        self.find_document(pipeline_id)
+            .and_then(|doc| doc.find_iframe(browsing_context_id))
+    }
+
+    pub fn iter(&self) -> DocumentsIter<'_> {
+        DocumentsIter {
+            iter: self.map.iter(),
+        }
+    }
+
+    /// Return the documents managed by this [`crate::script_thread::ScriptThread`] in the
+    /// order specified by the *[update the rendering][update-the-rendering]* step of the
+    /// HTML specification:
+    ///
+    /// > Let docs be all fully active Document objects whose relevant agent's event loop is
+    /// > eventLoop, sorted arbitrarily except that the following conditions must be met:
+    /// >
+    /// > Any Document B whose container document is A must be listed after A in the list.
+    /// >
+    /// > If there are two documents A and B that both have the same non-null container
+    /// > document C, then the order of A and B in the list must match the shadow-including
+    /// > tree order of their respective navigable containers in C's node tree.
+    /// >
+    /// > In the steps below that iterate over docs, each Document must be processed in the
+    /// > order it is found in the list.
+    ///
+    /// [update-the-rendering]: https://html.spec.whatwg.org/multipage/#update-the-rendering
+    pub(crate) fn documents_in_order(&self) -> Vec<PipelineId> {
+        // TODO: This is a fairly expensive operation, because iterating iframes requires walking
+        // the *entire* DOM for a document. Instead this should be cached and marked as dirty when
+        // the DOM of a document changes or documents are added or removed from our set.
+        DocumentTree::new(self).documents_in_order()
+    }
+}
+
+impl Default for DocumentCollection {
+    #[allow(crown::unrooted_must_root)]
+    fn default() -> Self {
+        Self {
+            map: HashMapTracedValues::new(),
+        }
+    }
+}
+
+#[allow(crown::unrooted_must_root)]
+pub struct DocumentsIter<'a> {
+    iter: hash_map::Iter<'a, PipelineId, Dom<Document>>,
+}
+
+impl<'a> Iterator for DocumentsIter<'a> {
+    type Item = (PipelineId, DomRoot<Document>);
+
+    fn next(&mut self) -> Option<(PipelineId, DomRoot<Document>)> {
+        self.iter
+            .next()
+            .map(|(id, doc)| (*id, DomRoot::from_ref(&**doc)))
+    }
+}
+
+#[derive(Default)]
+struct DocumentTreeNode {
+    parent: Option<PipelineId>,
+    children: Vec<PipelineId>,
+}
+
+/// A tree representation of [`Document`]s managed by the [`ScriptThread`][st], which is used
+/// to generate an ordered set of [`Document`]s for the *update the rendering* step of the
+/// HTML5 specification.
+///
+/// FIXME: The [`ScriptThread`][st] only has a view of [`Document`]s managed by itself,
+/// so if there are interceding iframes managed by other `ScriptThread`s, then the
+/// order of the [`Document`]s may not be correct. Perhaps the Constellation could
+/// ensure that every [`ScriptThread`][st] has the full view of the frame tree.
```

Comment:
In the "interceding iframes" case, does the ordering actually matter from the perspective of "update the rendering"? Or where we filtering out documents and not updating their rendering?

Reply:
Hrm. It's unclear to me if it matters or not. I'm not sure what kind of behavior the WPT suites are testing and if they are testing complex situations involving iframes nested with different origins. Exactly following the spec here would require a lot of tricky synchronization between different `ScriptThread`s (and potentially different processes) though.